src/integrator.jl

```diff
@@ -5,37 +5,76 @@
+jitter(::AbstractLeapfrog, ϵ) = ϵ
+temper_first(::AbstractLeapfrog, r, ::Int, ::Int) = r
+temper_second(::AbstractLeapfrog, r, ::Int, ::Int) = r
 
-# TODO: double check the function below to see if it is type stable or not
 function step(
-    lf::Leapfrog{F},
+    lf::AbstractLeapfrog{T},
     h::Hamiltonian,
     z::PhasePoint,
     n_steps::Int=1;
-    fwd::Bool=n_steps > 0 # simulate hamiltonian backward when n_steps < 0,
-) where {F<:AbstractFloat,T<:Real}
-    @unpack θ, r = z
+    fwd::Bool=n_steps > 0   # simulate hamiltonian backward when n_steps < 0,
+) where {T<:AbstractFloat}
+    n_steps = abs(n_steps)  # to support `n_steps < 0` cases
     ϵ = fwd ? lf.ϵ : -lf.ϵ
-    # Jitter step size; ref: https://github.com/stan-dev/stan/blob/1bb054027b01326e66ec610e95ef9b2a60aa6bec/src/stan/mcmc/hmc/base_hmc.hpp#L177-L178
-    iszero(lf.jitter) || (ϵ *= (1 + lf.jitter * (2 * rand() - 1)))
+    ϵ = jitter(lf, ϵ)
 
+    @unpack θ, r = z
     @unpack value, gradient = ∂H∂θ(h, θ)
-    for i = 1:abs(n_steps)
-        r = r - ϵ / 2 * gradient # take a half leapfrog step for momentum variable
+    for i = 1:n_steps
+        r = temper_first(lf, r, i, n_steps)
+        r = r - ϵ / 2 * gradient    # take a half leapfrog step for momentum variable
         ∇r = ∂H∂r(h, r)
-        θ = θ + ϵ * ∇r   # take a full leapfrog step for position variable
+        θ = θ + ϵ * ∇r              # take a full leapfrog step for position variable
         @unpack value, gradient = ∂H∂θ(h, θ)
-        r = r - ϵ / 2 * gradient # take a half leapfrog step for momentum variable
+        r = r - ϵ / 2 * gradient    # take a half leapfrog step for momentum variable
+        r = temper_second(lf, r, i, n_steps)
         # Create a new phase point by caching the logdensity and gradient
         z = phasepoint(h, θ, r; ℓπ=DualValue(value, gradient))
         !isfinite(z) && break
     end
     return z
 end
+
+struct Leapfrog{T<:AbstractFloat} <: AbstractLeapfrog{T}
+    ϵ       ::  T
+end
+Base.show(io::IO, l::Leapfrog) = print(io, "Leapfrog(ϵ=$(round(l.ϵ; sigdigits=3)))")
+
+### Jittering
+
+struct JitteredLeapfrog{T<:AbstractFloat} <: AbstractLeapfrog{T}
+    ϵ       ::  T
+    jitter  ::  T
+end
+
+function Base.show(io::IO, l::JitteredLeapfrog)
+    print(io, "JitteredLeapfrog(ϵ=$(round(l.ϵ; sigdigits=3)), jitter=$(round(l.jitter; sigdigits=3)))")
+end
+
+# Jitter step size; ref: https://github.com/stan-dev/stan/blob/1bb054027b01326e66ec610e95ef9b2a60aa6bec/src/stan/mcmc/hmc/base_hmc.hpp#L177-L178
+jitter(lf::JitteredLeapfrog, ϵ) = ϵ * (1 + lf.jitter * (2 * rand() - 1))
+
+### Tempering
+
+struct TemperedLeapfrog{T<:AbstractFloat} <: AbstractLeapfrog{T}
+    ϵ       ::  T
+    α       ::  T
+end
+
+function Base.show(io::IO, l::TemperedLeapfrog)
+    print(io, "TemperedLeapfrog(ϵ=$(round(l.ϵ; sigdigits=3)), α=$(round(l.α; sigdigits=3)))")
+end
+
+function temper_first(lf::TemperedLeapfrog, r, i::Int, n_steps::Int)
```

Comment:
Maybe merge `temper_first` and `temper_second` into `termper`? Since their function signature is the same, and we can distinguish multiplication and division by inspecting `i` and `n`.

Reply:
I'd like to keep the current syntax for now. If we merge them we just get a function with a if statement. Plus the calling is less readable (one cannot tell what does `true` mean in `temper(lf, r, i, n_steps ,true)` directly).

Reply:
That's true. Maybe move the odd-checking logic for leapfrog steps out of `temper`? E.g. 

```julia

temper(...)
# some stuff
!isodd(n_steps) && temper(...)
```

This way the `temper` function will also be simplified (not more odd-checking code).

Reply:
The code above has a bug. It adds an extra multiplication to moment if the number of leapfrog steps is odd. But maybe we can do the following 

```julia
!isodd(n_steps) && temper(...)
# some stuff
!isodd(n_steps) && temper(...)
```

This won't do any change to the moment for the `mid` leapfrog step if `n_step` is odd.

Reply:
By `isodd` do you mean `< div(n_steps, 2)` and `> div(n_steps, 2)`?

Reply:
```
!isodd(n_steps) && temper(...)
# some stuff
!isodd(n_steps) && temper(...)
``
Doesn't quite make sense to me. We always do this two `temper` steps regardless if `n_steps` is odd. The logic about the odd number thing is to implement Neal's tempering scheme in which the middle lf step do multiplication first then division. (Other lf steps either do 2 multiplication or 2 division).

Reply:
A slightly better way is to define

```julia

@enum TemperedState TemperStep1 TemperStep2
function temper(lf::TemperedLeapfrog, r, i::Int, n_steps::Int, s::TemperedState) end

# Then, 
temper(..., TemperStep1)
# some stuff
temper(..., TemperStep2)
```

Reply:
> By isodd do you mean < div(n_steps, 2) and > div(n_steps, 2)?

Julia has a built-in function `isodd`, e.g. `isodd(n_steps)`. 

> Doesn't quite make sense to me. We always do this two `temper` steps regardless if `n_steps` is odd. The logic about the odd number thing is to implement Neal's tempering scheme in which the middle lf step do multiplication first then division. (Other lf steps either do 2 multiplication or 2 division). 

Yes, there are different variants of the tempering trick, e.g. do nothing for the mid-step if `n_step` is odd.

Reply:
OK I see it. I can live with this but the current implementation also handles `isodd(n_steps)=true` correctly for Neal's version, what do we want to change?

Also even with the `isodd` function, we still need to have something like `temper_first` and `temper_second` or `@enum TemperedState TemperStep1 TemperStep2` to resolve which half it is. In fact, I don't like `@enum` here because we cannot dispatch over that, and this means we sill need a branch in that function, right?

I cannot see why
```julia
# Then, 
temper(..., TemperStep1)
# some stuff
temper(..., TemperStep2)
```
with
```julia
function temper(..., t)
    if t == TemperStep1
        ...
    elseif t == TemperStep2
        ...
    end
end
```
is over
```julia
# Then, 
temper_first(...)
# some stuff
temper_second(...)
```

Reply:
I think a bit more and feel your proposal of tempering (do nothing to odd) might be neater. I also think this scheme is correct. I will give a try tomorrow.

Reply:
@yebai I actually realize using iterators here would lead to much nicer code. Please check the update.

Reply:
Hi Kai, I feel the new code is a bit harder to read than before, so I did a bit refactoring of the tempering code without functionality change. It's basically changing the argument `i::Int` into ` tuple `(i::Int, ::Bool)`, where the bool value indicates first or second half of the leapfrog step.

Reply:
Fair enough. 

The code you wrote seems to be buggy. `false` was passed to `temper` in the first call but `if i[2] == true # first half of leapfrog` was written in `temper`, which means `i[2]` was used to indicate if the step is the first half tempering step, right?

In fact, I don't think using a tuple is readable 
- For the `step` function, passing `(i,false)` is mysterious - what does `false` mean here?
- For the `temper` function - `i[2] == true` means nothing to someone is reading only that function. 

I updated the current version to use a named tuple, fix and simplify the logic in `temper`.
The trick of `floor` and `ceil` I used was a bit mysterious TBH. In fact, the logic could be much simpler (as in my previous version):

Let `i` being the current leapfrog iteration counter and let `is_first` being whether or not it's the first step in the current leapfrog iteration. Then `i_temper`, the counter for half tempering step, is simply `2(i - 1) + 1 + is_first` and the logic for tempering is simply
- `if i_temper <= n_steps`, temper up
- `else`, temper down

This deals with both even and odd `n_steps`.

Reply:
Looks good. Maybe rename is_first to is_half? Since it indicates half and full leapfrog step here.

Reply:
Done